app/models/target.rb

```diff
@@ -21,13 +22,36 @@ class Target < ApplicationRecord
   belongs_to :topic
   belongs_to :user
 
+  has_one :match_users_conversation, dependent: :destroy
+  has_one :conversation, through: :match_users_conversation
+
   validates :title, presence: true
   validates :radius, presence: true, numericality: { greater_than: 0 }
   validates :lat, :lng, presence: true, numericality: true
-  validate  :max_target_amount_reached, unless: -> { user.nil? }
+  validate  :max_target_amount_reached, unless: -> { user.nil? }, on: :create
 
   MAX_TARGETS_AMOUNT = ENV.fetch('TARGET_CREATION_LIMIT', '3').to_i
 
+  scope :by_topic, ->(topic) { where(topic:) }
+  scope :not_same_user, ->(user) { where.not(user:) }
+  scope :inside_their_radius, lambda { |lng, lat|
```

Comment:
Wow that's interesting, this is postgresql syntax? 👏 

Another possibility would have been to use [geokit](https://github.com/geokit/geokit-rails) gem, that helps you on geolocation calculations